### PR TITLE
Generic vec implementation using Felt252Dict and Nullable

### PR DESCRIPTION
**Pull request by maciejka**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

- Fetl252Vec - for types that fits into a Felt252
- NullableVec - for types that don't fit in a Felt252

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

File: alexandria/data_structures/src/tests/vec_test.cairo [hunk 1]
```diff
@@ -1,106 +1,300 @@
...
-use array::ArrayTrait;
 use traits::{Index, Into, TryInto};
 use option::OptionTrait;
 use result::ResultTrait;
-use dict::Felt252DictTrait;
 
 // Internal imports
 use alexandria_data_structures::vec::VecTrait;
+use alexandria_data_structures::vec::Felt252Vec;
+use alexandria_data_structures::vec::NullableVec;
```
Comment (gaetbout) [inline alexandria/data_structures/src/tests/vec_test.cairo:9]:
```suggestion
use alexandria_data_structures::vec::{VecTrait, Felt252Vec, NullableVec};
```

File: alexandria/data_structures/src/vec.cairo [hunk 2]
```diff
@@ -81,42 +108,70 @@ impl VecImpl<
...
+        if index < self.len() {
+            Option::Some(self.items.get(index.into()).deref())
+        } else {
+            Option::None(())
+        }
+    }
+
+    fn at(ref self: NullableVec<T>, index: usize) -> T {
+        if index < self.len() {
+            return self.items.get(index.into()).deref();
```
Comment (gaetbout) [inline alexandria/data_structures/src/vec.cairo:155]:
```suggestion
           self.items.get(index.into()).deref()
```

**Review by gaetbout — APPROVED** [review]
Besides small changes LGTM

File: alexandria/data_structures/src/vec.cairo [hunk 3]
```diff
@@ -81,42 +108,70 @@ impl VecImpl<
...
+    fn set(ref self: NullableVec<T>, index: usize, value: T) {
+        if index < self.len() {
+            self.items.insert(index.into(), nullable_from_box(BoxTrait::new(value)));
+        } else {
+            panic_with_felt252('Index out of bounds')
+        }
+    }
+
+    fn len(self: @NullableVec<T>) -> usize {
+        *(self.len)
```
Comment (LucasLvy) [inline alexandria/data_structures/src/vec.cairo:175]: are the `()` needed ?

File: alexandria/data_structures/src/vec.cairo [hunk 4]
```diff
@@ -81,42 +108,70 @@ impl VecImpl<
...
+        self.items.insert(self.len.into(), nullable_from_box(BoxTrait::new(value)));
+        self.len = integer::u32_wrapping_add(self.len, 1_usize);
+    }
+
+    fn set(ref self: NullableVec<T>, index: usize, value: T) {
+        if index < self.len() {
+            self.items.insert(index.into(), nullable_from_box(BoxTrait::new(value)));
+        } else {
+            panic_with_felt252('Index out of bounds')
+        }
```
Comment (LucasLvy) [inline alexandria/data_structures/src/vec.cairo:171]: assert then insert

File: alexandria/data_structures/src/vec.cairo [hunk 5]
```diff
@@ -81,42 +108,70 @@ impl VecImpl<
...
+            Option::None(())
+        }
+    }
+
+    fn at(ref self: NullableVec<T>, index: usize) -> T {
+        if index < self.len() {
+            return self.items.get(index.into()).deref();
+        } else {
+            panic_with_felt252('Index out of bounds')
+        }
```
Comment (LucasLvy) [inline alexandria/data_structures/src/vec.cairo:158]: assert then return

**Review by LucasLvy — APPROVED** [review]

**Pull request merged**